webtau-groovy/src/main/groovy/com/twosigma/webtau/cli/interactive/WebTauCliInteractive.groovy

```diff
@@ -0,0 +1,278 @@
+
+        readAndHandleIdxOrCommand { idx ->
+            if (idx < 0 || idx >= scenarios.size()) {
+                ConsoleOutputs.out(Color.RED, 'enter scenario index')
+                return false
+            } else {
+                selectedScenarioIdx = idx
+                state = State.ScenarioActionSelection
+                return true
+            }
+        }
+    }
+
+    void selectScenarioAction() {
+        def scenarioCommands = commandsForCurrentState()
+        displaySelectedScenario()
+        displayCommands(scenarioCommands)
+
+        readAndHandleIdxOrCommand { idx -> return false }
+    }
+
+    void displaySelectedScenario() {
+        def filePath = testFilePaths[selectedPathIdx]
+        def tests = testByPath[filePath]
+        def test = tests[selectedScenarioIdx]
+
+        ConsoleOutputs.out(Color.PURPLE, filePath, ': ', Color.YELLOW, test.scenario)
+    }
+
+    void runSelectedScenario() {
+        def filePath = testFilePaths[selectedPathIdx]
+        def tests = testByPath[filePath]
+
+        def test = tests[selectedScenarioIdx]
+        test.run()
+
+        if (test.exception) {
+            ConsoleOutputs.out(Color.RED, StackTraceUtils.renderStackTraceWithoutLibCalls(test.exception))
+        }
+
+        state = State.ScenarioActionSelection
+    }
+
+    IdxOrCommand readIdxOrCommand() {
+        def allowedCommands = commandsForCurrentState()
+
+        while (true) {
+            def line = readLine()
+            def idxOrCommand = new IdxOrCommand(line)
+
+            def command = idxOrCommand.command
+
+            if ((!command || allowedCommands.find { it.matches() == command }) && idxOrCommand.idx == null) {
+                ConsoleOutputs.out(Color.RED, 'enter a number or a command')
+                displayCommands(allowedCommands)
+            } else {
+                return idxOrCommand
+            }
+        }
+    }
+
+    String readLine() {
+        print new AutoResetAnsiString(Color.GREEN, 'webtau', Color.YELLOW, ' > ')
+        return inReader.readLine()
+    }
+
+    void readAndHandleIdxOrCommand(Closure idxHandler) {
+        while (true) {
+            def idxOrCommand = readIdxOrCommand()
+
+            if (idxOrCommand.command) {
+                handleCommand(idxOrCommand.command)
+                return
+            }
+
+            def done = idxHandler(idxOrCommand.idx)
+            if (done) {
+                return
+            }
+        }
+    }
+
+    private handleCommand(InteractiveCommand command) {
+        switch (command) {
+            case InteractiveCommand.Back:
+                handleBackCommand()
+                break
+            case InteractiveCommand.Quit:
+                state = State.Done
+                break
+            case InteractiveCommand.Run:
+                runSelectedScenario()
+                break
+        }
+    }
+
+    void handleBackCommand() {
+        switch (state) {
```

Comment:
Should we unset relevant idx's or are we ok relying on the state transition machinery to make sure we're not accidentally relying on those?

Reply:
that part is being refactored at the moment. But basically it is to contain a latest user selection